qiling/extensions/r2/r2.py

```diff
@@ -0,0 +1,227 @@
+        self._cmd(f'wx {code.hex()}')
+        # set architecture and bits for r2 asm
+        self._cmd(f"e,asm.arch={self.ql.arch.type.name.lower().removesuffix('64')},asm.bits={self.ql.arch.bits}")
+
+    def _setup_file(self, path: str):
+        path = path.encode()
+        fh = libr.r_core.r_core_file_open(self._r2c, path, UC_PROT_READ | UC_PROT_EXEC, self.loadaddr)
+        libr.r_core.r_core_bin_load(self._r2c, path, self.baseaddr)
+
+    def _cmd(self, cmd: str) -> str:
+        r = libr.r_core.r_core_cmd_str(
+            self._r2c, ctypes.create_string_buffer(cmd.encode("utf-8")))
+        return ctypes.string_at(r).decode('utf-8')
+
+    @staticmethod
+    def aaa(fun):
+        @wraps(fun)
+        def wrapper(self):
+            if self.analyzed is False:
+                self._cmd("aaa")
+                self.analyzed = True
+            return fun(self)
+        return wrapper
+
+    def _cmdj(self, cmd: str) -> Union[Dict, List[Dict]]:
+        return json.loads(self._cmd(cmd))
+    
+    def read(self, addr: int, size: int) -> bytes:
+        hexstr = self._cmd(f"p8 {size} @ {addr}")
+        return bytes.fromhex(hexstr)
+
+    @cached_property
+    def sections(self) -> Dict[str, Section]:
+        sec_lst = self._cmdj("iSj")
+        return {dic['name']: Section(**dic) for dic in sec_lst}
+
+    @cached_property
+    def strings(self) -> Dict[str, String]:
+        str_lst = self._cmdj("izzj")
+        return {dic['string']: String(**dic) for dic in str_lst}
+
+    @cached_property
+    def symbols(self) -> Dict[str, Symbol]:
+        sym_lst = self._cmdj("isj")
+        return {dic['name']: Symbol(**dic).vaddr for dic in sym_lst}
+
+    @cached_property
+    @aaa
+    def functions(self) -> Dict[str, Function]:
+        fcn_lst = self._cmdj("aflj")
+        return {dic['name']: Function(**dic) for dic in fcn_lst}
+
+    @cached_property
+    @aaa
+    def flags(self) -> List[Flag]:
+        return [Flag(**dic) for dic in self._cmdj("fj")]
+
+    @cached_property
+    @aaa
+    def xrefs(self) -> Dict[int, Xref]:
+        return {dic['from']: Xref(**dic) for dic in self._cmdj("axj")}
+
+    def at(self, addr: int) -> Tuple[Flag, int]:
+        # the most suitable flag should have address <= addr
+        # bisect_right find the insertion point, right side if value exists
+        idx = bisect.bisect_right(self.flags, Flag(offset=addr))
+        # minus 1 to find the corresponding flag
+        flag = self.flags[idx - 1]
+        return flag, addr - flag.offset
+
+    def refrom(self, addr: int) -> Optional[Xref]:
+        return self.xrefs.get(addr)
+
+    def refto(self, addr: int) -> List[Xref]:
+        return [xref for xref in self.xrefs.values() if xref.addr == addr]
+
+    def enable_trace(self, mode='full'):
+        # simple map from addr to flag name, cannot resolve addresses in the middle
+        self.ql.loader.symsmap = {flag.offset: flag.name for flag in self.flags}
+        if mode == 'full':
```

Comment:
Use an enumeration instead.

Reply:
This is just a PoC, the trace extension made by @elicn need a `symsmap` to resolve addresses, I have to made further modification on the trace extension.